### PR TITLE
Transactions

### PR DESCRIPTION
**Pull request by alexholmes99**
Users are able to transfer money between each other, and view these transactions.

### CODE REVIEW

File: app/controllers/transactions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                    render 'new'
+                end
+            else
+                flash[:notice] = "User does not have this account"
+                render 'new'
+            end
+        else
+            flash[:notice] = "User does not exist"
+            render 'new'
+        end
```
Comment (wosborne) [inline app/controllers/transactions_controller.rb]: This has a lot of logic in it and I think we can reduce and make this nicer to read by extracting all the checks and flash notices into their own methods.

So we could make the whole action look like
```
  def create
        @transaction = Transaction.new(transaction_params)
        @account = @user.accounts.find(@transaction.sender_id)
        @receiver = find_receiver
        @receiver_act = find_receiver_act
        if account_has_sufficient_funds && @transaction.save
           @account.update(:balance => (@account.balance - @transaction.amount))
           @receiver_act.update(:balance => (@receiver_act.balance + @transaction.amount))
           flash[:success] = "Transaction Completed"
           redirect_to user_url(@user)
        else
           flash[:error] = @transaction.errors.full_messages
           render 'new'
        end
    end
```

Comment (wosborne) [reply app/controllers/transactions_controller.rb]: then for example the `find_receiver` can be a private method and look like
```
def find_receiver
  receiver = User.find_by(username: @transaction.receiver_name)
  if receiver.present?
    return receiver
  else
    flash[:notice] = "User does not exist"
    render 'new'
  end
end
```

Comment (wosborne) [reply app/controllers/transactions_controller.rb]: or a slightly shorter version
```
def find_receiver
  receiver = User.find_by(username: @transaction.receiver_name)
  return receiver unless receiver.present?
  flash[:notice] = "User does not exist"
  render 'new'
end
```

File: app/controllers/transactions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    private
+    def find_user
+        @user = User.find(params[:user_id])
+    end
+
+    private
+    def transaction_params
+        params.require(:transaction).permit(:transaction_id, :accounts_id, :users_id, :sender_id, :sender_name, :receiver_name, :receiver_act, :amount, :reference)
+    end
```
Comment (wosborne) [inline app/controllers/transactions_controller.rb]: `private` makes all methods underneath private. you can remove the one on line 46

File: app/models/transaction.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Transaction < ApplicationRecord
+    belongs_to :account, optional: true
```
Comment (wosborne) [inline app/models/transaction.rb:2]: does a transaction not always belong_to an account ?

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -1,17 +1,20 @@
...
+    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.1.3/dist/js/bootstrap.bundle.min.js" integrity="sha384-ka7Sk0Gln4gmtz2MlQnikT1wXgYsOg+OMhuP+IlRH9sENBO0LRn5q+8nbTov4+1p" crossorigin="anonymous"></script>
+    <script src='//cdn.datatables.net/1.11.3/css/jquery.dataTables.min.css'></script>
+    <script src='//cdn.datatables.net/1.11.3/js/jquery.dataTables.min.js'></script>
     <title>BankApp</title>
     <meta name="viewport" content="width=device-width,initial-scale=1">
     <%= csrf_meta_tags %>
     <%= csp_meta_tag %>
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
     <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
-  </head>
```
Comment (wosborne) [inline app/views/layouts/application.html.erb:14]: put this head closing tag back :)

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -1,17 +1,20 @@
...
+    
+    
 
 
     <nav class="navbar navbar-expand-lg navbar-light">
```
Comment (wosborne) [inline app/views/layouts/application.html.erb:24]: Think `<nav>` should be within `<body>`

File: package.json [hunk 6]
```diff
@@ -6,6 +6,7 @@
     "@rails/activestorage": "^6.0.0",
     "@rails/ujs": "^6.0.0",
     "@rails/webpacker": "5.4.3",
+    "expo-cli": "^4.12.5",
```
Comment (wosborne) [inline package.json]: can remove this, snuck in during react native setup

File: config/routes.rb [hunk 7]
```diff
@@ -4,12 +4,13 @@
...
-    resources :accounts
+    resources :transactions
+    resources :accounts 
   end
-
   #Session routes
   get '/login', to: 'sessions#login'
   post '/login', to: 'sessions#create'
+  get '/logout', to: 'sessions#destroy'
   delete '/logout', to: 'sessions#destroy'
```
Comment (wosborne) [inline config/routes.rb]: dont think you need both of these, if the `delete` works then remove `get` and visa versa

File: config/routes.rb [hunk 8]
```diff
@@ -4,12 +4,13 @@
   get 'unauthorised' => 'pages#unauthorised'
   #Users routes
   resources :users do
-    resources :accounts
+    resources :transactions
+    resources :accounts 
   end
-
   #Session routes
   get '/login', to: 'sessions#login'
```
Comment (wosborne) [inline config/routes.rb:11]: I don't see a `def login` in your sessions controller so i think this line can be removed.

**Review by wosborne — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 8 above)
Comment (alexholmes99) [reply config/routes.rb:11]: I think its all to do with a gem I'm using, when I delete either login route my code fails

**Review by wosborne — APPROVED** [review]

**Pull request merged**